src/librustc_middle/traits/mod.rs

```diff
@@ -80,8 +81,39 @@ pub enum Reveal {
 }
 
 /// The reason why we incurred this obligation; used for error reporting.
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+///
+/// As the happy path does not care about this struct, storing this on the heap
+/// ends up increasing performance.
+///
+/// We do not want to intern this as there are a lot of obligation causes which
+/// only live for a short period of time.
+#[derive(Clone, PartialEq, Eq, Hash)]
 pub struct ObligationCause<'tcx> {
+    /// `None` for `ObligationCause::dummy`, `Some` otherwise.
+    data: Option<Rc<ObligationCauseData<'tcx>>>,
+}
+
+// Correctly format `ObligationCause::dummy`.
+impl<'tcx> fmt::Debug for ObligationCause<'tcx> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        ObligationCauseData::fmt(self, f)
+    }
+}
+
+impl Deref for ObligationCause<'tcx> {
+    type Target = ObligationCauseData<'tcx>;
+
+    fn deref(&self) -> &Self::Target {
+        self.data.as_deref().unwrap_or(&ObligationCauseData {
+            span: DUMMY_SP,
+            body_id: hir::CRATE_HIR_ID,
+            code: MiscObligation,
+        })
```

Comment:
Sweet! Sorry for the churn, but would you mind using a single associated `const` for the dummy `ObligationCauseData` here and in `make_mut`?

r=me with that done. Do you want to go back in the perf queue before this is merged or just go ahead since this will be rollup=never anyways?

Reply:
Done.

Let's merge this without checking perf again. As we use `rollup=never` this should be fine.